BGP CP: Adds Support for Multi-Pool IPAM

danehans opened the pull request:
Adds Support for Multi-Pool IPAM:

- Adds `PodIPPoolSelector` to `CiliumBGPVirtualRouter` for defining a label selector for advertising routes from allocated IPAM pools.
- Adds `LocalCiliumNodeResource ` to the BGP CP `Controller` to provide a stream of events for changes to the local CiliumNode resource.
- Adds `PodIPPoolAnnouncements` to `ServerWithConfig` for holding all the announced multi-pool IPAM CIDRs.
- Adds `PodIPPoolReconciler` to advertise IPAM pool CIDRs of a matching CiliumBGPPeeringPolicy.
- Updates BGP CP unit and integration tests.

Fixes: #26937

```release-note
The podIPPoolSelector field has been added to CiliumBGPVirtualRouter for selectively advertising multi-pool IPAM CIDRs.
```

rastislavs commented on pkg/bgpv1/manager/reconcile_test.go:
```diff
@@ -546,6 +546,231 @@ func TestExportPodCIDRReconciler(t *testing.T) {
+			},
+			advertised: []netip.Prefix{
+				netip.MustParsePrefix("192.168.0.0/24"),
+				//netip.MustParsePrefix("192.168.3.0/24"),
```
why is this commented out?

danehans replied:
Commit https://github.com/cilium/cilium/pull/27100/commits/36751e473cd16d8777a456ea34d8f5f20a8d62b2 fixes this.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -218,6 +229,16 @@ type CiliumBGPVirtualRouter struct {
 
+// IPPoolLabelSelector selects multi-pool IPAM CIDRs that have been allocated to the
+// CiliumNode(s) matching this policy.
+type IPPoolLabelSelector struct {
```
__Note:__ This selector type was added since the [MatchLabelsValue](https://pkg.go.dev/github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1#MatchLabelsValue)  of a `slimv1.LabelSelector` considers an IP prefix invalid due to:
```
+kubebuilder:validation:Pattern=`^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?$`
```

rastislavs replied:
Hmm, I guess the issue here is that we are not really matching by labels (`metadata.labels` on `CiliumPodIPPool` CRs) which I think was the original idea [in the issue](https://github.com/cilium/cilium/issues/26937#issuecomment-1647671516), but by pool name + CIDRs.

Not sure what is the reasoning/motivation here - do you think that matching whole pools would not be granular enough? I think this deviates a bit from the other selectors that we have (`NodeSelector` / `ServiceSelector` both match labels of respective resources).

That said, if we go this way, I don't think we should name this `IPPoolLabelSelector` - probably just `IPPoolSelector`, and the map should not be named `MatchLabels` - as the labels are not really being matched.

danehans replied:
> Hmm, I guess the issue here is that we are not really matching by labels (metadata.labels on CiliumPodIPPool CRs) which I think was the original idea in the issue, but by pool name + CIDRs.

The name of this type and the type's field does not properly reflect the implementation. I will update the PR accordingly.

> Not sure what is the reasoning/motivation here - do you think that matching whole pools would not be granular enough? I think this deviates a bit from the other selectors that we have (NodeSelector / ServiceSelector both match labels of respective resources).

IMO the pool name is not granular enough and may cause unintended advertisements when CIDRs are added to a pool. I was using the [issue](https://github.com/cilium/cilium/issues/26937#issuecomment-1647671516) for implementation guidance. I do not have a strong opinion on the granularity of pool advertisements, so I'll defer to maintainers on how to proceed.

danehans replied:
Commit https://github.com/cilium/cilium/commit/709a5fa23cec5b99988d6560896b8c436a9a57f1 renames the API type and `matchLabel` field.

YutaroHayakawa replied:
While I agree that sometimes the granularity is sometimes not enough, I'm not a big fan of this idea, because 1) since users can change the CIDR on PodIPPool, it's very easy to introduce configuration drift between BGP CPlane and PodIPPool (e.g. users update PodIPPool, but forget to update BGP CPlane config) 2) it's not consistent with existing behavior with ClusterPoolV2Beta (we advertise all prefixes provided).

I'd solve this problem by introducing mechanisms like Cisco's routemap (it doesn't solve the problem of configuration drift though)

danehans replied:
@YutaroHayakawa thanks for the review. Commit https://github.com/cilium/cilium/commit/caac36cfeb1658128bd3b03636cc2b0c264a318e implements your comments above.

YutaroHayakawa commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -199,6 +199,13 @@ type CiliumBGPVirtualRouter struct {
+	// If unspecified, no IPAM pool CIDRs will be advertised.
+	//
+	// +kubebuilder:validation:Optional
+	PodIPPools []string `json:"podIPPools,omitempty"`
```
Do you have any reason you ended up to list PodIPPools instead of selecting them by labels?

danehans replied:
It seems more efficient not having to Get/List a CiliumPodIPPool when a CiliumNode or CiliumBGPPeeringPolicy changes. IPAM assigns a CIDR from the pool to a CiliumNode, storing the latest copy in the local node store. This causes observers, e.g. BGP CP, to trigger a reconciliation using CiliumBGPPeeringPolicy and CiliumNode (local node store). Thoughts?

YutaroHayakawa replied:
From an efficiency point of view, that makes sense to me. My only concern is from a UX point of view, I'm not sure it's comfortable for users to need to enumerate all pools they want to advertise one by one. For example, let's say there's a user who wants to advertise all pools. In that case, they need to edit two resources when they add a new pool. With label-based selection, they just need to label the new pool and they don't need to edit BGPPeeringPolicy. From this perspective, I'd prefer label-based selection. WDYT?

danehans replied:
@YutaroHayakawa thanks for the additional feedback. I can see how providing a list of pools can be a suboptimal UX when managing a large number of pools. On the other hand, the current implementation is more efficient. Is it worth providing both options or do you think the UX of label-based selection outweighs the efficiency of the current approach? Here's an example of supporting both selection types:
```go
type NodeSpec struct {
...
	// PodIPPools advertises CIDRS from CilimPodIPPools that have been allocated to the
	// CiliumNodes matching this policy.
	PodIPPools *PodIPPools
}

type PodIPPools struct {
	// NameSelector advertises CilimPodIPPools based on pools names. Ignored if LabelSelector is specified
	NameSelector []string

	// LabelSelector advertises CiliumPodIPPools based on labels....
	LabelSelector *slimv1.LabelSelector
}
```

We could initially omit `NameSelector` but still use `PodIPPools` to support other selection types in the future. Thoughts?

YutaroHayakawa replied:
Sounds nice, but sorry I changed my mind and thought your current implementation is better. Thinking well about the label-based design, I realized that it requires all agents to watch `CiliumPodIPPool`, which is not the case right now (currently, only Cilium Operator watches `CiliumPodIPPool`). This adds significant churn to API servers, especially in the large cluster, so I think the cost exceeds the benefit.

danehans replied:
@YutaroHayakawa thanks for looking into various approaches to the implementation and providing feedback. I thought of another approach to support label-based `CiliumPodIPPool` selection. We could extend [IPAMPoolAllocation](https://github.com/cilium/cilium/blob/main/pkg/ipam/types/types.go#L59-L71) to include `CiliumPodIPPool` labels. The IPAM allocator would then write the labels from `CiliumPodIPPool` to the `CiliumNode`. BGP CP could use a LabelSelector from the CiliumBGPPeeringPolicy to select CIDRs for announcing based on the newly added labels. Thoughts?

rastislavs replied:
@YutaroHayakawa 
>  I realized that it requires all agents to watch CiliumPodIPPool ... . This adds significant churn to API servers,

Is it really an issue? Each agent is already connected to the apiserver, and this just adds one more type to watch. I would not expect many updates happening on the `CiliumPodIPPool` CRs, IMO usually users set it once and would not touch that it even for months.

I just realized a similar issue could be then with my PR as well, which now needs agents to watch for `CiliumLoadBalancerIPPool`: https://github.com/cilium/cilium/pull/27705

YutaroHayakawa replied:
> We could extend [IPAMPoolAllocation](https://github.com/cilium/cilium/blob/main/pkg/ipam/types/types.go#L59-L71) to include CiliumPodIPPool labels.

Yeah, if we want to avoid watching `CiliumPodIPPool`, that would be the way.

> Is it really an issue? Each agent is already connected to the apiserver, and this just adds one more type to watch. I would not expect many updates happening on the CiliumPodIPPool CRs, IMO usually users set it once and would not touch that it even for months.

Umm, I think you're right here. In the case of `CiliumPodIPPool`'s use case, as you said, it might not be a big issue. Since I don't fully understand the use cases of `CiliumPodIPPool` I'd like to hear @gandro 's opinion here.

danehans replied:
> We could extend [IPAMPoolAllocation](https://github.com/cilium/cilium/blob/main/pkg/ipam/types/types.go#L59-L71) to include CiliumPodIPPool labels.

[Here](https://github.com/danehans/cilium/tree/issue_26937_labels) is a WIP branch for implementing the label-based approach. I'm having issues updating the CiliumNode IPAM allocation pool labels based on CiliumPodIPPool labels but I think the implementation is close.

gandro replied:
> We could extend IPAMPoolAllocation to include CiliumPodIPPool labels. The IPAM allocator would then write the labels from CiliumPodIPPool to the CiliumNode.

I worry a bit that this will cause synchronization bugs, because we're now duplicating mutable information across CRDs. If we do this, it means we have to update all CiliumNode objects when a CiliumPodIPPool label changes, which is about as costly as just watching all CiliumPodIPPool from each agent. So I'm not sure what we will save by doing this.

It's too early to know all the different use cases of CiliumPodIPPools, but I expect there not to be a more than a few dozen per cluster. So I don't expect watching all pools from all agents to be too costly overall.

YutaroHayakawa reviewed (changes requested):
I did an initial go through. Overall looks great to me!

Let's cleanup commits and add proper commit messages to describe the changes. I think first, second, and the last commits can be consolidated. Third and forth commits also can be consolidated?

YutaroHayakawa commented on pkg/ipam/multipool.go:
```diff
@@ -495,6 +499,10 @@ func (m *multiPoolManager) updateCiliumNode(ctx context.Context) error {
+		// Update the pools in local node store.
+		m.localNodeStore.Update(func(n *node.LocalNode) {
+			n.IPAMAllocPools = newNode.AllocatedIPAMPools()
+		})
```
Q: Why do we need this? Above, we update the CiliumNode resource, so someone catch that event and update LocalNodeStore no?

danehans replied:
> someone catch that event and update LocalNodeStore

That is not happening (verified by setting logging throughout). This is needed since https://github.com/cilium/cilium/pull/27285 updated BGP CP to use the LocalNodeStore.

YutaroHayakawa commented on pkg/node/types/node.go:
```diff
@@ -395,6 +427,14 @@ func (n *Node) GetCiliumInternalIP(ipv6 bool) net.IP {
 	return nil
 }
 
+// GetIPAMAllocPools...
```
Maybe this comment is WIP?

danehans replied:
Ack. Fixed in the latest commit.

YutaroHayakawa reviewed (changes requested):
Did a second look. Thanks for the commit fix! However, could you separate IPAM-related changes and BGP-related changes into different commits? Currently, it's hard to follow what is the bug fix and what is a new feature.

Also, while I was testing the changes on my local Kind environment, changing `virtualRouters[].podIPPools` doesn't trigger reconciliation. 

Repro

1. Deploy Cilium with this configuration

```
ipam:
  mode: multi-pool
  operator:
    autoCreateCiliumPodIPPools:
      default:
        ipv4:
          cidrs:
            - 10.10.0.0/16
          maskSize: 24
      other:
        ipv4:
          cidrs:
            - 10.20.0.0/16
          maskSize: 24
bgpControlPlane:
  enabled: true
tunnel: disabled
ipv4NativeRoutingCIDR: <your native routing CIDR>
autoDirectNodeRoutes: true
```

```
helm install cilium -n kube-system oci://quay.io/cilium-charts-dev/cilium --version 1.15.0-dev-dev.703-HEAD-823c7b6751 -f values.yaml
```

2. Deploy daemonset with `other` pool selected

```
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: netshoot
spec:
  selector:
    matchLabels:
      app: netshoot
  template:
    metadata:
      labels:
        app: netshoot
      annotations:
        ipam.cilium.io/ip-pool: other
    spec:
      tolerations:
      - key: node-role.kubernetes.io/master
        operator: Exists
        effect: NoSchedule
      containers:
      - name: netshoot
        image: nicolaka/netshoot:latest
        command: ["sleep", "infinite"]
```

3. Deploy BGPP

```
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPPeeringPolicy
metadata:
  name: test
spec:
  nodeSelector:
    matchLabels:
      kubernetes.io/hostname: kind-control-plane
  virtualRouters:
    - localASN: 65000
      podIPPools:
      - default
      neighbors:
        - peerASN: 65000
          peerAddress: <your peer>
```

4. Confirm the `default` prefix is advertised
5. Add `other` to `podIPPools`

In my case, 5 didn't update the BGP advertisement. The debug log shows `level=debug msg="No reconciliation necessary" component=manager.podIPPoolReconciler subsys=bgp-control-plane`. So, probably the reconciler doesn't recognize the diff.

gandro commented on pkg/ipam/ipam.go:
```diff
@@ -107,7 +108,13 @@ type Metadata interface {
+	k8sEventReg K8sEventRegister,
+	mtuConfig MtuConfiguration,
+	clientset client.Clientset,
+	nodeStore *node.LocalNodeStore) *IPAM {
```
I'm not a big fan of introducing this dependency here. 

In particular, we have https://github.com/cilium/cilium/pull/26962 which attempts to decouple some of the dependencies by having both IPAM and BGP rely on a `LocalCiliumNodeResource`.

Given that the list of allocated CIDRs is not coming from the agent IPAM module, but instead is coming from the operator, there is not really a need for this IPAM module to tell BGP or the local node store that we have a new CIDR. Instead, BGP can obtain this information directly from the `LocalCiliumNodeResource` I think, as we already do with the singular pod CIDRs in IPAM kubernetes/clusterpool mode.

danehans replied:
@gandro thanks for the review. I'll work on refactoring for BGP CP to use `LocalCiliumNodeResource` instead of introducing the `LocalNodeStore` dep.

danehans replied:
> In particular, we have https://github.com/cilium/cilium/pull/26962 which attempts to decouple some of the dependencies by having both IPAM and BGP rely on a LocalCiliumNodeResource.

^ PR decouples the deprecated BGP, e.g. MetalLB, and not BGP CP. https://github.com/cilium/cilium/pull/27285 introduced `LocalNodeStore` to BGP CP for node state so I was trying to follow this approach in commit https://github.com/cilium/cilium/pull/27100/commits/1b4ce953a443a7706bf3d0845c3c44f0d7e1f453.

> Instead, BGP can obtain this information directly from the LocalCiliumNodeResource I think, as we already do with the singular pod CIDRs in IPAM kubernetes/clusterpool mode.

I will explore this approach instead of adding `LocalNodeStore` to the IPAM module.

ldelossa replied:
@gandro https://github.com/cilium/cilium/pull/27285 moved to using LocalNodeStore to decouple ourselves from relying on a particular representation of the Node. My intention was to continue the use of LocalNodeStore to keep this decoupled approach. I was under the impression that this was the preferred method moving forward.

I understand this change is in the IPAM code, so your discretion should be advised but just wanted to give you context on why the BGP-CP took this approach.

danehans replied:
Unless I'm missing something, it seems like adding `LocalCiliumNodeResource` to BGP when `LocalNodeStore` already exists is duplicative.

danehans replied:
Commit 0329660 follows the recommendations of https://github.com/cilium/cilium/pull/27100#discussion_r1307355084 by using `LocalCiliumNodeResource` to trigger events instead of the internal `Node` type.

bimmlerd commented on pkg/node/types/node.go:
```diff
@@ -231,6 +260,9 @@ type Node struct {
 	IPv6SecondaryAllocCIDRs []*cidr.CIDR
 
+	// IPAMAllocPools ia a list of IPAM pools allocated from the local CiliumNode.
+	IPAMAllocPools []ipamTypes.IPAMPoolAllocation
```
Let's not add a field here. `node.Node` is de-facto stable API; it get's serialized to JSON into etcd, so removing fields is difficult. It's also what other nodes consume of this node, and I don't think the `IPAMAllocPools` of this node need to be sent to all other nodes.

Instead, how about adding this field to the `LocalNode` type in `pkg/node/local_node_store.go`.

danehans replied:
Same feedback as https://github.com/cilium/cilium/pull/27100#discussion_r1316676569.

bimmlerd commented on pkg/node/address.go:
```diff
@@ -357,6 +358,19 @@ func SetIPv4AllocRange(net *cidr.CIDR) {
 	})
 }
 
+func GetIPAMAllocPools() []ipamTypes.IPAMPoolAllocation {
```
I realise that this is entirely non-obvious, but these package global functions are historic, and we should avoid adding new ones. They are here for compatability reasons, while we transition the code to a more modular approach. See https://github.com/cilium/cilium/pull/21191 for a bit of history.

Instead, depend on the local node store and call update on it, if you need to set something.

danehans replied:
Let me look into the approach you are suggesting. I was following the changes from [this](https://github.com/cilium/cilium/pull/27285/files#diff-6cb81d56efa78ad58fc0dd7038341c90aad7382fe7fd160282f6fec7ab014ca3R26-R31) recent PR.

danehans replied:
Commit 0329660 follows the recommendations of https://github.com/cilium/cilium/pull/27100#discussion_r1307355084 by using `LocalCiliumNodeResource` to trigger events instead of the internal `Node` type.

bimmlerd commented on pkg/k8s/init.go:
```diff
@@ -153,6 +163,8 @@ func WaitForNodeInformation(ctx context.Context, k8sGetter k8sGetter) error {
 			logfields.K8sNodeIP:        k8sNodeIP,
 		}).Info("Received own node information from API server")
 
+		useNodeIPAMAllocPools(n)
```
Would it be possible to write the reconciler in a way that we don't have to do this here? As in, do we strictly depend on having this information available early on, and if so, why?

danehans replied:
> do we strictly depend on having this information available early on

I don't think the allocation pools are needed this early. Let me test and I'll update the PR or provide additional feedback.

danehans replied:
Commit 0329660 removed this change.

bimmlerd reviewed (changes requested):
Thanks for the ping @gandro, and thanks for your contribution @danehans.

I've taken a look, and tried to comment in specific places what things I'd like to see done differently.

danehans committed:
BGP CP: Adds Support for Multi-Pool IPAM

- Adds `PodIPPoolSelector` to `CiliumBGPVirtualRouter` for defining
  a label selector for advertising routes from allocated IPAM pools.
- Adds `LocalCiliumNodeResource ` to the BGP CP `Controller` to provide
  a stream of events for changes to the local CiliumNode resource.
- Adds `PodIPPoolAnnouncements` to `ServerWithConfig` for holding all
  the announced multi-pool IPAM CIDRs.
- Adds `PodIPPoolReconciler` to advertise IPAM pool CIDRs of a matching
  CiliumBGPPeeringPolicy.
- Updates BGP CP unit and integration tests.

Signed-off-by: Daneyon Hansen <daneyon.hansen@solo.io>

YutaroHayakawa commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -381,6 +385,210 @@ func (r *ExportPodCIDRReconciler) Reconcile(ctx context.Context, p ReconcilePara
+func (r *PodIPPoolReconciler) Reconcile(ctx context.Context, p ReconcileParams) error {
+	lp := r.populateLocalPools(p.CiliumNode)
+
+	if err := r.fullReconciliation(ctx, p.CurrentServer, p.DesiredConfig, lp); err != nil {
```
Nit: If you always do full reconciliation, you don't need to use DiffStore and can use Store directly.

rastislavs replied:
There is a bit of inconvenience by using the Store directly in the reconciler (need to write the code for event signalling as well) - that is why I introduced the `BGPCPResourceStore` here: https://github.com/cilium/cilium/pull/27705/commits/5e8c7c8cdf3406de5b3b89a0d721cf7ff8397fc5 . When that PR (https://github.com/cilium/cilium/pull/27705) is merged, we can just replace the `Diffstore` with the new `BGPCPResourceStore` in this reconciler as well.

danehans replied:
@rastislavs thanks for the heads-up. I'll start working on rebasing this PR on top of https://github.com/cilium/cilium/pull/27705 and use `BGPCPResourceStore`.

youngnick commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -112,15 +137,23 @@ func NewController(params ControllerParams) (*Controller, error) {
+	if err != nil {
+		return fmt.Errorf("IPPoolResource.Store(): %w", err)
+	}
+	c.IPPooolLister = ipPoolListerFunc(func() ([]*v2alpha1api.CiliumPodIPPool, error) {
```
`IPPooolLister` doesn't seem right?

danehans replied:
Thanks for the review @youngnick. Commit b18c075 implements this change request.

youngnick reviewed (approved):
LGTM aside from one possible typo?

bimmlerd commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -74,13 +95,15 @@ type Controller struct {
+	IPPoolResource       resource.Resource[*v2alpha1api.CiliumPodIPPool]
+	PolicyResource       resource.Resource[*v2alpha1api.CiliumBGPPeeringPolicy]
+	DaemonConfig         *option.DaemonConfig
+	LocalCiliumNodeStore daemon_k8s.LocalCiliumNodeResource
```
```suggestion
	LocalCiliumNodeResource daemon_k8s.LocalCiliumNodeResource
```
probably?

danehans replied:
Commit b18c075 implements this change request.

bimmlerd commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -134,6 +167,32 @@ func (c *Controller) Start(startCtx hive.HookContext) error {
+	c.workerpool.Submit("cilium-node-observer", func(ctx context.Context) error {
+		for ev := range c.CiliumNodeResource.Events(ctx) {
+			switch ev.Kind {
+			case resource.Upsert, resource.Delete:
```
I think it's unexpected to receive a delete for the local CiliumNode. Is setting the LocalCiliumNode to `nil` really what we want in this case? AFAIR, other code just ignores the delete (possibly logging a warning, but the warning doesn't seem all that actionable by a user, so let's not) and continues on with the last known CiliumNode.

danehans replied:
@bimmlerd I agree that a CiliumNode delete is unexpected. However, BGP CP would continue to advertise the pod IP pools in this case. Are there any valid use cases for deleting/recreating a CiliumNode? @rastislavs @YutaroHayakawa thoughts?

YutaroHayakawa replied:
I think deleting `CiliumNode` is generally invalid. In that sense, ignoring the `CiliumNode` deletion and keeping the advertisement is not a bad option.

danehans replied:
@bimmlerd @YutaroHayakawa thanks for the guidance here. Commit 3181c3a ignores `CiliumNode` deletion events.

YutaroHayakawa reviewed (approved):
LGTM with one nit 👍  Thanks again for your contribution! I also did manual testing on my local, and it seems it is working as expected.

I don't have any objection to removing LocalNodeStore and using LocalCiliumNodeStore. As we saw in this PR, we need to manually define the update trigger to get the right value, and for me, watching the CiliumNode resource directly looks much simpler. Do you have any other opinion @ldelossa ?

bimmlerd reviewed (commented):
I'm okay with the way this interacts with nodes now, thanks. couple of comments, but starting to look good.

gandro commented on pkg/bgpv1/cell.go:
```diff
@@ -54,3 +58,19 @@ func newBGPPeeringPolicyResource(lc hive.Lifecycle, c client.Clientset, dc *opti
+		return nil
+	}
+
+	return resource.New[*v2alpha1api.CiliumPodIPPool](
```
My understanding is that this will start a K8s watcher for CiliumPodIPPools. Since they are only needed if the IPAM mode is multi-pool, could we return `nil` if `IPAMMode != MultiPool`?

danehans replied:
Commit b18c075 implements this change request.

gandro commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -108,7 +109,10 @@ func NewBGPRouterManager(params bgpRouterManagerParams) agent.BGPRouterManager {
-func (m *BGPRouterManager) ConfigurePeers(ctx context.Context, policy *v2alpha1api.CiliumBGPPeeringPolicy, node *node.LocalNode) error {
+func (m *BGPRouterManager) ConfigurePeers(ctx context.Context,
+	policy *v2alpha1api.CiliumBGPPeeringPolicy,
+	node *node.LocalNode,
+	ciliumNode *v2api.CiliumNode) error {
```
nit (feel free to ignore): Would it make sense to introduce a new `localNode` struct that contains both the `node.LocalNode` as well as the `CiliumNode`? It feels a bit weird that we're basically passing in the same node object twice.

bimmlerd replied:
Might even be worth adding a reference local cilium node to the LocalNode type in the LocalNodeStore - @jussi wdyt?

danehans replied:
> nit (feel free to ignore): Would it make sense to introduce a new localNode struct that contains both the node.LocalNode as well as the CiliumNode? It feels a bit weird that we're basically passing in the same node object twice.

I agree but the [plan](https://github.com/cilium/cilium/pull/27100#issuecomment-1707611384) is to remove the `node.LocalNode` and just roll with the `LocalCiliumNode` resource after this PR merges.

gandro commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -381,6 +385,210 @@ func (r *ExportPodCIDRReconciler) Reconcile(ctx context.Context, p ReconcilePara
+			if p, err := cidr.ToPrefix(); err == nil && (p.Addr().Is4() || p.Addr().Is6()) {
+				prefixes = append(prefixes, *p)
+			} else {
+				log.Errorf("invalid ipam pool cidr: %v for cilium node %s", p, localNode.Name)
```
nit: Consider using `WithFields` when logging such values.

I think it could also be useful to log `err` here too

danehans replied:
Commit b18c075 implements this change request.

gandro commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -381,6 +385,210 @@ func (r *ExportPodCIDRReconciler) Reconcile(ctx context.Context, p ReconcilePara
+	for _, pool := range localNode.Spec.IPAM.Pools.Allocated {
+		var prefixes []netip.Prefix
+		for _, cidr := range pool.CIDRs {
+			if p, err := cidr.ToPrefix(); err == nil && (p.Addr().Is4() || p.Addr().Is6()) {
```
Out of curiosity: Why is the `(p.Addr().Is4() || p.Addr().Is6())` check needed? Is there case where an IP is neither?

danehans replied:
Thanks for catching this. Yes, these checks are unneeded and removed in commit b18c075.

gandro commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -381,6 +385,210 @@ func (r *ExportPodCIDRReconciler) Reconcile(ctx context.Context, p ReconcilePara
+
+	for _, desiredRoute := range desiredRoutes {
+		// If this route has already been announced, don't add it again
+		if slices.IndexFunc(sc.PodIPPoolAnnouncements[poolKey], func(existing *types.Path) bool {
```
nit: Wouldn't `slices.ContainsFunc` be a better match here?

danehans replied:
Commit b18c075 implements this change request.

gandro commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -381,6 +385,210 @@ func (r *ExportPodCIDRReconciler) Reconcile(ctx context.Context, p ReconcilePara
+	for i := len(sc.PodIPPoolAnnouncements[poolKey]) - 1; i >= 0; i-- {
+		announcement := sc.PodIPPoolAnnouncements[poolKey][i]
+		// If the announcement is within the list of desired routes, don't remove it
+		if slices.IndexFunc(desiredRoutes, func(existing netip.Prefix) bool {
```
Same as above regarding `slices.ContainsFunc`

danehans replied:
Commit b18c075 implements this change request.

gandro commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -199,6 +199,13 @@ type CiliumBGPVirtualRouter struct {
+	// PodIPPoolSelector selects CiliumPodIPPools based on labels. The virtual
+	// router will announce allocated CIDRs of matching CiliumPodIPPools.
+	//
+	// If empty / nil no CiliumPodIPPools will be announced.
```
Hm. How does one announce all pools? Does this require that every pool has a label?

Given that the service selector also has this behavior, it's probably a sensible default, but I'm still curious how to do that. Because I expect that it will be very common that users want to announce all pools that a node has pod IPs allocated from.

danehans replied:
Yes, the `PodIPPoolSelector` is consistent with the `ServiceSelector` by only selecting pools with matching labels (and allocated to the local CiliumNode).

> I expect that it will be very common that users want to announce all pools that a node has pod IPs allocated from.

We can reshape the API to expose other selection options, e.g. all pools from the cluster or a particular namespace. Since the shape and behavior of `PodIPPoolSelector` is consistent with `ServiceSelector`, both fields should be updated to meet this use case. This is already a large PR and these are alpha API changes, so I prefer to support this use case in a follow-on PR. If so, I can create an issue and assign myself.

christarazi replied:
In K8s, typically an empty selector means it selects all. A nil selector selects nothing. At least the former is how Cilium's CNP `endpointSelector` works. The reason the latter is not implemented is because CNPs cannot have a nil `endpointSelector`. 

But 

* empty => all
* nil => nil

makes sense to me. Thoughts @cilium/sig-k8s?

christarazi replied:
Given that the CRD is alpha, we don't have to be extra careful about breaking the API, so this is ok with me for now. However, I think we need to have a larger discussion on how these selectors behave differently across resources and wider compared to K8s. Could you bring this up during the next community meeting?

gandro commented on pkg/k8s/init.go:
```diff
@@ -64,7 +64,9 @@ func retrieveNodeInformation(ctx context.Context, nodeGetter k8sGetter, nodeName
-	if option.Config.IPAM == ipamOption.IPAMClusterPool || option.Config.IPAM == ipamOption.IPAMClusterPoolV2 {
+	if option.Config.IPAM == ipamOption.IPAMClusterPool ||
+		option.Config.IPAM == ipamOption.IPAMClusterPoolV2 ||
+		option.Config.IPAM == ipamOption.IPAMMultiPool {
```
Why is this needed? MultiPool should neither require nor set a local alloc CIDR.

danehans replied:
@gandro thanks for the review and catching this leftover change. Now that this PR no longer uses the `Node` type to store pod IP pools, this change is unneeded. I will remove it.

gandro reviewed (commented):
Thanks a lot, this is great work! Overall the BGP logic looks good to me, though I'm not an expert.

I've left some smaller feedback on a few things that stood out to me

christarazi commented on pkg/ipam/types/types.go:
```diff
@@ -56,6 +57,18 @@ type AllocationMap map[string]AllocationIP
+		return &prefix, nil
+	}
+
+	return nil, fmt.Errorf("failed to parse ipam cidr %v", c)
```
The error is not reported in the return here.

danehans replied:
@christarazi thanks for the review! Commit d6c79cf resolves this issue.

nathanjsweet reviewed (approved):
LGTM for API.

rastislavs commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -44,12 +46,31 @@ func (plf policyListerFunc) List() ([]*v2alpha1api.CiliumBGPPeeringPolicy, error
+	// to the objects in its store.
+	IPPoolResource resource.Resource[*v2alpha1api.CiliumPodIPPool]
+	// IPPoolLister is an interface which allows for the listing of the CiliumPodIPPool resource.
+	IPPoolLister ipPoolLister
```
Do we still need `IPPoolResource` and `IPPoolLister` in the controller now that we use `BGPCPResourceStore[*v2alpha1api.CiliumPodIPPool]` in the reconciler?

danehans replied:
Both are unneeded and have been removed in commit 572b94b.

rastislavs commented on pkg/bgpv1/manager/pod_ip_pool_reconciler.go:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	// Loop over all pod ip pools, reconcile any updates to the pool.
+	iter := r.poolStore.IterKeys()
```
Nit: you can also use `r.poolStore.List()` instead of the iterator

danehans replied:
Commit 572b94b uses the returned value from `r.poolStore.List()` as the iterator.

rastislavs reviewed (approved):
LGTM now, thanks!

christarazi reviewed (approved):
LGTM for my codeowners

bimmlerd reviewed (approved):
Since maintainers-little-helper insists :roll_eyes: 
LGTM
